src/main/scala/generatortools/io/Utils.scala

```diff
@@ -0,0 +1,156 @@
+            FixedPoint(128.W, 64.BP)
+          case (UnknownWidth(), KnownBinaryPoint(bp), _) if bp > 0 =>
+            // Make a large FP placeholder when W is unknown & fractional bits are used
+            FixedPoint((64 + bp).W, bp.BP)
+          case (UnknownWidth(), KnownBinaryPoint(bp), lower: IsKnown) if bp == 0 && lower.value >= 0 =>
+            // Make large UInt when W is unknown, but i is an integer with lower bound >= 0
+            UInt(64.W)
+          case (UnknownWidth(), KnownBinaryPoint(bp), _) if bp == 0 =>
+            // Make large SInt when W is unknown and lower bound is unknown
+            SInt(64.W)
+          case (KnownWidth(width), KnownBinaryPoint(bp), lower: IsKnown) if bp > 0 =>
+            // Equivalent FixedPoint representation
+            FixedPoint(width.W, bp.BP)
+          case (KnownWidth(width), KnownBinaryPoint(bp), lower: IsKnown) if bp == 0 && lower.value < 0 =>
+            // Equivalent SInt representation
+            SInt(width.W)
+          case (KnownWidth(width), KnownBinaryPoint(bp), lower: IsKnown) if bp == 0 && lower.value >= 0 =>
+            // Equivalent UInt representation -- Interval is always signed
+            UInt((width - 1).W)
+          case (_, _, _) =>
+            throw new Exception(s"Invalid type to convert $tpe")
+        }
+    }
+    // Match data direction
+    if (CheckDirection.isInput(tpe)) Input(newTpe)
+    else if (CheckDirection.isOutput(tpe)) Output(newTpe)
+    else newTpe
+  }
+}
+
+object CheckDirection {
+  /** Is data Input? */
+  def isInput(d: Data): Boolean = matchDirection(d)._1
+  /** Is data Output? */
+  def isOutput(d: Data): Boolean = matchDirection(d)._2
+  /** Checks if Element is input, output, or neither. Aggregates are always neither. */
+  private def matchDirection(d: Data): (Boolean, Boolean) = d match {
+    // TODO: Make consistent with new Chisel
+    case e: Element =>
+      DataMirror.directionOf(e) match {
+        case chisel3.core.ActualDirection.Input => (true, false)
+        case chisel3.core.ActualDirection.Output => (false, true)
+        case _ => (false, false)
+      }
+    // Fake base type
+    case r: DspReal =>
+      DataMirror.directionOf(r) match {
+        case chisel3.core.ActualDirection.Input => (true, false)
+        case chisel3.core.ActualDirection.Output => (false, true)
+        case _ => (false, false)
+      }
+    case c: DspComplex[_] =>
+      val (in1, out1) = matchDirection(c.real)
+      val (in2, out2) = matchDirection(c.imag)
+      (in1 & in2, out1 & out2)
+    case _ => (false, false)
+  }
+}
+
+object Connect {
+  /** Connect similar signals. Hierarchy should be identical (esp. for Records). This enables you to, for example,
+    * connect an Interval signal with an equivalent base Chisel signal. globalClk will be passed in to
+    * all clock inputs, if provided.
+    */
+  def apply[T <: Data, U <: Data](left: T, right: U, globalClk: Option[Clock] = None): Unit = {
```

Comment:
This also seems a bit scary, and in this PR is only used in one place. Might be better to just do the connection by hand if there aren't other usages.